pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -360,62 +387,89 @@ func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachine
+	}
+	return nil
+}
+
+func (c *DisruptionBudgetController) createPDB(key string, vmi *virtv1.VirtualMachineInstance) error {
+	minAvailable := intstr.FromInt(1)
+
+	c.podDisruptionBudgetExpectations.ExpectCreations(key, 1)
+	createdPDB, err := c.clientset.PolicyV1beta1().PodDisruptionBudgets(vmi.Namespace).Create(context.Background(), &v1beta1.PodDisruptionBudget{
+		ObjectMeta: v1.ObjectMeta{
+			OwnerReferences: []v1.OwnerReference{
+				*v1.NewControllerRef(vmi, virtv1.VirtualMachineInstanceGroupVersionKind),
 			},
-			Spec: v1beta1.PodDisruptionBudgetSpec{
-				MinAvailable: &two,
-				Selector: &v1.LabelSelector{
-					MatchLabels: map[string]string{
-						virtv1.CreatedByLabel: string(vmi.UID),
-					},
+			GenerateName: "kubevirt-disruption-budget-",
+		},
+		Spec: v1beta1.PodDisruptionBudgetSpec{
+			MinAvailable: &minAvailable,
+			Selector: &v1.LabelSelector{
+				MatchLabels: map[string]string{
+					virtv1.CreatedByLabel: string(vmi.UID),
 				},
 			},
-		}, v1.CreateOptions{})
-		if err != nil {
-			c.podDisruptionBudgetExpectations.CreationObserved(key)
-			c.recorder.Eventf(vmi, v12.EventTypeWarning, FailedCreatePodDisruptionBudgetReason, "Error creating a PodDisruptionBudget: %v", err)
-			return err
-		}
-		c.recorder.Eventf(vmi, v12.EventTypeNormal, SuccessfulCreatePodDisruptionBudgetReason, "Created PodDisruptionBudget %s", createdPDB.Name)
+		},
+	}, v1.CreateOptions{})
+	if err != nil {
+		c.podDisruptionBudgetExpectations.CreationObserved(key)
+		c.recorder.Eventf(vmi, corev1.EventTypeWarning, FailedCreatePodDisruptionBudgetReason, "Error creating a PodDisruptionBudget: %v", err)
+		return err
 	}
+	c.recorder.Eventf(vmi, corev1.EventTypeNormal, SuccessfulCreatePodDisruptionBudgetReason, "Created PodDisruptionBudget %s", createdPDB.Name)
 	return nil
 }
 
-func (c *DisruptionBudgetController) pdbForVMI(namespace, name string) (*v1beta1.PodDisruptionBudget, error) {
+func isPDBFromOldVMI(vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) bool {
+	// The pdb might be from an old vmi with a different uid, delete and later create the correct one
+	// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
+	if pdb == nil {
+		return false
+	}
+	ownerRef := v1.GetControllerOf(pdb)
+	return ownerRef != nil && ownerRef.UID != vmi.UID
+}
+
+func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) error {
+	migratableOnDrain := vmiMigratableOnDrain(vmi)
+
+	if vmi == nil || vmi.DeletionTimestamp != nil ||
+		!migratableOnDrain || (migratableOnDrain && isPDBFromOldVMI(vmi, pdb)) {
+		return c.deletePDB(key, pdb, vmi)
+	}
+
+	if pdb != nil {
+		// only delete the temporary migration PDB if migration is complete
+		return c.deleteOnVMIMCompletion(key, pdb, vmi)
+	}
+	// No pdb and we want migrations on evictions
+	return c.createPDB(key, vmi)
+}
```

Comment:
I think what you have here is technically correct. from a code quality/readability/maintainability standpoint here's a suggestion for how maybe the code could be organized in a way that's more approachable. 

What do you think? I know it's more code, but it's clear exactly what conditions lead to what action. 

```

create := false
delete := false
migratableOnDrain := vmiMigratableOnDrain(vmi)

// check for deletions if pod exists
if pdb != nil {
      if vmi == nil || vmi.DeletionTimestamp != nil{
          // being deleted
          delete = true
      } else if !migratableOnDrain {
          // vmi isn't set to migrate on eviction, so delete. 
          delete = true
      } else if isPDBFromOldVMI(vmi, pdb) {
          // pdb for non existent vmi
          delete = true
      } else if isVMIMCompletedForPDB(key, pdb, vmi) {
          // pdb for completed migration
          delete = true
      }
} else if migratableOnDrain {
    // pdb doesn't exist, create if vmi is set to migrate during drain. 
    create = true
}


if create {
    return c.createPDB(key, vmi)
} else if delete {
    return c.deletePDB(key, pdb, vmi)
}

return nil
```

Reply:
yup it makes sense, I'll just keep the 'return early statements' instead of the additional create and delete variables if that's ok.